### PR TITLE
Refactor logic to allow to draw 256 banners at the same time

### PR DESCRIPTION
**Pull request by ZehMatt**
From this:

To this:

### CODE REVIEW

**Comment by Gymnasiast** [comment]
@LordMarcel Can you check if this solves your problem in #9631 ?

**Comment by ZehMatt** [comment]
I suspect this is the same issue

File: src/openrct2/drawing/Drawing.Sprite.cpp [hunk 1]
```diff
@@ -180,6 +180,7 @@ void mask_scalar(
 static rct_gx _g1 = {};
 static rct_gx _g2 = {};
 static rct_gx _csg = {};
+static rct_g1_element _scrollingText[256]{};
```
Comment (Gymnasiast) [inline src/openrct2/drawing/Drawing.Sprite.cpp:183]: Shouldn't this be
```suggestion
static rct_g1_element _scrollingText[MAX_SCROLLING_TEXT_ENTRIES]{};
```
?

File: src/openrct2/drawing/ScrollingText.cpp [hunk 2]
```diff
@@ -34,7 +34,7 @@ struct rct_draw_scroll_text
     uint8_t bitmap[64 * 40];
 };
 
-constexpr int32_t MAX_SCROLLING_TEXT_ENTRIES = 32;
+constexpr int32_t MAX_SCROLLING_TEXT_ENTRIES = SPR_SCROLLING_TEXT_END - SPR_SCROLLING_TEXT_START;
```
Comment (Gymnasiast) [inline src/openrct2/drawing/ScrollingText.cpp:37]: Normally, we would set this to a fixed value and define `SPR_SCROLLING_TEXT_END` as `SPR_SCROLLING_TEXT_START + 256`.

File: src/openrct2/drawing/Drawing.Sprite.cpp (hunk 1 above)
Comment (ZehMatt) [reply src/openrct2/drawing/Drawing.Sprite.cpp:183]: That constant is defined in ScrollingText.cpp

File: src/openrct2/drawing/ScrollingText.cpp (hunk 2 above)
Comment (ZehMatt) [reply src/openrct2/drawing/ScrollingText.cpp:37]: I agree, however there is no header file specific for banners/scrolling text, Its a bit out of place to define it in sprites.h so I left it alone.

File: src/openrct2/drawing/Drawing.Sprite.cpp (hunk 1 above)
Comment (Gymnasiast) [reply src/openrct2/drawing/Drawing.Sprite.cpp:183]: That's no good reason to hardcode it here. It completely defeats the point of having a constant in the first place.

Comment (ZehMatt) [reply src/openrct2/drawing/Drawing.Sprite.cpp:183]: Where do you suggest I define the constant? It would be required in sprites.h and ScrollingText.cpp

Comment (Gymnasiast) [reply src/openrct2/drawing/Drawing.Sprite.cpp:183]: I suggest we ask Ted or Duncan. I would like one of them to review the PR anyway.

Comment (ZehMatt) [reply src/openrct2/drawing/Drawing.Sprite.cpp]: I've introduced a header file called GameLimits.h, been wanting to do that for a while now. I think this simplifies such cases and makes it easier to control the compilation out of one place.

File: src/openrct2/drawing/ScrollingText.cpp [hunk 3]
```diff
@@ -97,10 +96,11 @@ void scrolling_text_initialise_bitmaps()
...
     }
 
-    for (int32_t i = 0; i < MAX_SCROLLING_TEXT_ENTRIES; i++)
+    for (int32_t i = 0; i < OpenRCT2::MaxScrollingTextEntries; i++)
     {
-        int32_t imageId = SPR_SCROLLING_TEXT_START + i;
-        const rct_g1_element* g1original = gfx_get_g1_element(imageId);
+        const int32_t imageIdReference = SPR_SCROLLING_TEXT_LEGACY_START;
+        const int32_t imageId = SPR_SCROLLING_TEXT_START + i;
+        const rct_g1_element* g1original = gfx_get_g1_element(imageIdReference);
```
Comment (duncanspumpkin) [inline src/openrct2/drawing/ScrollingText.cpp:103]: This is always getting the same element why is it in the loop? Remove it from the loop and also its nullptr check. I don't quite understand why you are doing it like this. Is there some specific reason you get a different g1 element overwrite most of its elements and then place it in an image location? Wouldn't it make more sense to build a new g1 element and then write it to the `imageId`?

**Comment by duncanspumpkin** [comment]
For my understanding of reading this you are removing the limit by ignoring the built in 32 entries and making a new 256 entry patch at a different location. Is that correct?

**Comment by ZehMatt** [comment]
> For my understanding of reading this you are removing the limit by ignoring the built in 32 entries and making a new 256 entry patch at a different location. Is that correct?

Yes, I use the 256 ids before the dynamic image table, I wasn't sure if using it anywhere else would be safe.

File: src/openrct2/drawing/ScrollingText.cpp (hunk 3 above)
Comment (ZehMatt) [reply src/openrct2/drawing/ScrollingText.cpp:103]: Not copying the original sprite(s) caused black artifacts, but yes having it in the loop is pointless I just need the data as reference.

Comment (duncanspumpkin) [reply src/openrct2/drawing/ScrollingText.cpp:103]: Might as well work out what the other fields you are copying are and remove the requirement to copy

File: src/openrct2/GameLimits.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/*****************************************************************************
```
Comment (Gymnasiast) [inline src/openrct2/GameLimits.h:1]: Calling it GameLimits.h is too broad - there are dozens, if not hundreds of such limits. I would call it `src/openrct2/drawing/ScrollingText.h`.

Comment (ZehMatt) [reply src/openrct2/GameLimits.h:1]: The purpose of this file is to be precisely that, broad. We should move all technical limits in there, right now its way too scattered.

File: src/openrct2/drawing/ScrollingText.cpp [hunk 5]
```diff
@@ -98,23 +98,25 @@ void scrolling_text_initialise_bitmaps()
...
-            gfx_set_g1_element(imageId, &g1);
-        }
+
+        // Initialize the scrolling text sprite.
+        rct_g1_element g1{};
+        g1.offset = _drawScrollTextList[i].bitmap;
+        g1.x_offset = -32;
+        g1.y_offset = 0;
+        g1.flags = G1_FLAG_BMP;
+        g1.zoomed_offset = 766;
```
Comment (duncanspumpkin) [inline src/openrct2/drawing/ScrollingText.cpp:109]: Define an spr constant for this

File: src/openrct2/GameLimits.h (hunk 4 above)
Comment (Gymnasiast) [reply src/openrct2/GameLimits.h:1]: I have to disagree here .What we’ll end up with is a mega file that gets impossible to read. There are hundreds of limits.

Comment (ZehMatt) [reply src/openrct2/GameLimits.h:1]: If there are so many limits its even more reason to have a central location for them and not have hundreds of constants scattered in my opinion.

**Comment by IntelOrca** [comment]
You are reducing 256 images for objects to use now. You might be able to get a much larger limit by swapping out a single allocated g2 entry with a different dynamic set before drawing each banner.

File: src/openrct2/GameLimits.h (hunk 4 above)
Comment (Gymnasiast) [reply src/openrct2/GameLimits.h:1]: It will only complicate matters and make every part of the code tied to it IMO.

**Review by Gymnasiast — CHANGES_REQUESTED** [review]
Please change the GameLimits.h file into ScrollingText.h.

**Comment by ZehMatt** [comment]
> You are reducing 256 images for objects to use now. You might be able to get a much larger limit by swapping out a single allocated g2 entry with a different dynamic set before drawing each banner.

The current limit of images is 0x7FFFF (524287) which is half a million, I don't think the 256 gonna hurt that much. Adding more complexity for something like this is not feasible in my opinion.

**Comment by duncanspumpkin** [comment]
I also think the reduction is not an issue we are decreasing the limit by way more than that in the RMC PR's anyway. We need to start thinking more about how we can move that limit up to 32bit anyway.

**Comment by IntelOrca** [comment]
> I also think the reduction is not an issue we are decreasing the limit by way more than that in the RMC PR's anyway. We need to start thinking more about how we can move that limit up to 32bit anyway.

We need everywhere to use ImageId so that we can increase the ID mask.

**Review by Gymnasiast — APPROVED** [review]
All my requests have been fulfilled. I don't know if taking away 256 images is acceptable or not, so I'll leave that discussion to you, Duncan and Ted.

**Review by IntelOrca — APPROVED** [review]

**Pull request merged**